File: src/ui/macos/QuickAccessToolbar.swift
import AppKit
import ExcelCore

class QuickAccessToolbar: NSView {
    // MARK: - Properties
    private var stackView: NSStackView
    private var defaultItems: [ToolbarItem] = []
    private var customItems: [ToolbarItem] = []
    private var currentWorkbook: ExcelCore.Workbook

    // MARK: - Initialization
    init(workbook: ExcelCore.Workbook) {
        self.currentWorkbook = workbook
        self.stackView = NSStackView()
        
        super.init(frame: .zero)
        
        setupUI()
        loadDefaultItems()
        loadCustomItems()
        configureEventHandlers()
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    // MARK: - UI Setup
    private func setupUI() {
        // Create and configure the NSStackView
        stackView = NSStackView(frame: bounds)
        stackView.orientation = .horizontal
        stackView.spacing = 4
        stackView.distribution = .fillEqually
        
        // Add default and custom items to the stackView
        for item in defaultItems + customItems {
            stackView.addArrangedSubview(item.button)
        }
        
        // Apply styling to match Excel's theme
        stackView.wantsLayer = true
        stackView.layer?.backgroundColor = NSColor.controlBackgroundColor.cgColor
        
        addSubview(stackView)
        
        // Set up constraints
        stackView.translatesAutoresizingMaskIntoConstraints = false
        NSLayoutConstraint.activate([
            stackView.topAnchor.constraint(equalTo: topAnchor),
            stackView.leadingAnchor.constraint(equalTo: leadingAnchor),
            stackView.trailingAnchor.constraint(equalTo: trailingAnchor),
            stackView.bottomAnchor.constraint(equalTo: bottomAnchor)
        ])
    }

    // MARK: - Loading Items
    private func loadDefaultItems() {
        // Create ToolbarItem instances for default actions
        let saveItem = ToolbarItem(title: "Save", image: NSImage(named: "save_icon"), action: #selector(handleItemClick(_:)))
        let undoItem = ToolbarItem(title: "Undo", image: NSImage(named: "undo_icon"), action: #selector(handleItemClick(_:)))
        let redoItem = ToolbarItem(title: "Redo", image: NSImage(named: "redo_icon"), action: #selector(handleItemClick(_:)))
        
        // Add default items to the defaultItems array
        defaultItems = [saveItem, undoItem, redoItem]
    }

    private func loadCustomItems() {
        // Retrieve custom toolbar items from user settings
        let userSettings = ExcelCore.UserSettings.shared
        let customItemsData = userSettings.customToolbarItems
        
        // Create ToolbarItem instances for each custom item
        for itemData in customItemsData {
            if let item = ToolbarItem(from: itemData) {
                customItems.append(item)
            }
        }
    }

    // MARK: - Event Handling
    private func configureEventHandlers() {
        for item in defaultItems + customItems {
            item.button.target = self
            item.button.action = #selector(handleItemClick(_:))
        }
    }

    @objc private func handleItemClick(_ sender: NSButton) {
        // Identify the clicked toolbar item
        guard let item = (defaultItems + customItems).first(where: { $0.button == sender }) else {
            return
        }
        
        // Execute the associated command
        do {
            try ExcelCore.Command.execute(item.command, workbook: currentWorkbook)
        } catch {
            // Handle error (e.g., show an alert)
            let alert = NSAlert()
            alert.messageText = "Error"
            alert.informativeText = "Failed to execute command: \(error.localizedDescription)"
            alert.alertStyle = .warning
            alert.addButton(withTitle: "OK")
            alert.runModal()
        }
        
        // Update the UI state if necessary
        updateUIState()
    }

    private func updateUIState() {
        // Update the state of toggleable items
        for item in defaultItems + customItems {
            if item.isToggleable {
                item.button.state = item.isEnabled ? .on : .off
            }
        }
    }

    // MARK: - Customization
    @objc private func customizeToolbar() {
        // Create and show the customization window
        let customizationWindow = QuickAccessToolbarCustomizationWindow(toolbar: self)
        customizationWindow.showWindow(nil)
    }
}

// MARK: - Human Tasks
/*
 TODO: Implement a customization dialog for adding/removing Quick Access Toolbar items
 TODO: Add support for keyboard shortcuts for Quick Access Toolbar items
 TODO: Implement proper state management for toggleable toolbar items
 TODO: Add support for grouping related toolbar items
 TODO: Implement tooltips with extended information for toolbar items
 TODO: Add support for displaying recently used commands in the Quick Access Toolbar
 TODO: Implement synchronization of Quick Access Toolbar settings across devices
 TODO: Add support for exporting and importing Quick Access Toolbar configurations
 TODO: Implement accessibility features for VoiceOver
 TODO: Add support for Touch Bar integration for Quick Access Toolbar items
 TODO: Ensure proper scaling of toolbar for different screen resolutions
 TODO: Implement customizable toolbar positioning (left/right alignment)
 TODO: Determine the most useful default items for the Quick Access Toolbar
 TODO: Implement localization for default item labels and tooltips
 TODO: Implement a user interface for managing custom Quick Access Toolbar items
 TODO: Ensure proper serialization and deserialization of custom toolbar items
 TODO: Implement proper error handling for failed command executions
 TODO: Add support for command parameters in toolbar items
 TODO: Design and implement a user-friendly customization interface
 TODO: Ensure proper handling of drag-and-drop operations for item reordering
*/

File: src/ui/windows/QuickAccessToolbar.cs
using System;
using System.Collections.Generic;
using System.Windows.Controls;
using System.Windows.Input;
using ExcelCore;

namespace create_excel_v2_dru9yl.src.ui.windows
{
    /// <summary>
    /// Represents the Quick Access Toolbar in Excel
    /// </summary>
    public class QuickAccessToolbar
    {
        private ToolBar toolbar;
        private List<ToolbarItem> defaultItems;
        private List<ToolbarItem> customItems;
        private ExcelCore.Workbook currentWorkbook;

        /// <summary>
        /// Initializes a new instance of the QuickAccessToolbar class
        /// </summary>
        /// <param name="workbook">The current workbook</param>
        public QuickAccessToolbar(ExcelCore.Workbook workbook)
        {
            // Initialize the toolbar control
            InitializeToolbar();

            // Set the current workbook
            currentWorkbook = workbook;

            // Load default toolbar items
            LoadDefaultItems();

            // Load custom toolbar items from user settings
            LoadCustomItems();

            // Populate the toolbar with items
            PopulateToolbar();
        }

        /// <summary>
        /// Initializes the Quick Access Toolbar control
        /// </summary>
        private void InitializeToolbar()
        {
            // Create a new ToolBar control
            toolbar = new ToolBar();

            // Set toolbar properties (e.g., orientation, overflow behavior)
            toolbar.Orientation = Orientation.Horizontal;
            toolbar.OverflowMode = OverflowMode.AsNeeded;

            // Apply styling to match Excel's theme
            toolbar.Style = (Style)Application.Current.Resources["ExcelQuickAccessToolbarStyle"];
        }

        /// <summary>
        /// Loads the default items for the Quick Access Toolbar
        /// </summary>
        private void LoadDefaultItems()
        {
            defaultItems = new List<ToolbarItem>();

            // Create ToolbarItem instances for default actions (e.g., Save, Undo, Redo)
            defaultItems.Add(new ToolbarItem("Save", new SaveCommand(currentWorkbook), "Save_Icon.png"));
            defaultItems.Add(new ToolbarItem("Undo", new UndoCommand(currentWorkbook), "Undo_Icon.png"));
            defaultItems.Add(new ToolbarItem("Redo", new RedoCommand(currentWorkbook), "Redo_Icon.png"));
        }

        /// <summary>
        /// Loads custom items for the Quick Access Toolbar from user settings
        /// </summary>
        private void LoadCustomItems()
        {
            customItems = new List<ToolbarItem>();

            // Retrieve custom toolbar items from user settings
            var userSettings = ExcelCore.UserSettings.GetCurrentUserSettings();
            var savedCustomItems = userSettings.GetQuickAccessToolbarItems();

            // Create ToolbarItem instances for each custom item
            foreach (var item in savedCustomItems)
            {
                var command = ExcelCore.CommandFactory.CreateCommand(item.CommandName, currentWorkbook);
                customItems.Add(new ToolbarItem(item.Label, command, item.IconPath));
            }
        }

        /// <summary>
        /// Populates the Quick Access Toolbar with default and custom items
        /// </summary>
        private void PopulateToolbar()
        {
            // Clear existing items from the toolbar
            toolbar.Items.Clear();

            // Add default items to the toolbar
            foreach (var item in defaultItems)
            {
                AddItemToToolbar(item);
            }

            // Add custom items to the toolbar
            foreach (var item in customItems)
            {
                AddItemToToolbar(item);
            }

            // Add a dropdown button for toolbar customization
            var customizeButton = new Button
            {
                Content = "▼",
                ToolTip = "Customize Quick Access Toolbar"
            };
            customizeButton.Click += (sender, e) => ShowCustomizationDialog();
            toolbar.Items.Add(customizeButton);
        }

        private void AddItemToToolbar(ToolbarItem item)
        {
            var button = new Button
            {
                Content = new Image { Source = new BitmapImage(new Uri(item.IconPath, UriKind.Relative)) },
                ToolTip = item.Label,
                Command = item.Command
            };
            toolbar.Items.Add(button);
        }

        /// <summary>
        /// Handles click events for toolbar items
        /// </summary>
        /// <param name="sender">The sender object</param>
        /// <param name="e">The event arguments</param>
        private void HandleItemClick(object sender, EventArgs e)
        {
            // Identify the clicked toolbar item
            var clickedButton = sender as Button;
            var toolbarItem = clickedButton.Tag as ToolbarItem;

            // Execute the associated command
            try
            {
                toolbarItem.Command.Execute(null);
            }
            catch (Exception ex)
            {
                // TODO: Implement proper error handling for failed command executions
                MessageBox.Show($"Error executing command: {ex.Message}", "Command Execution Error", MessageBoxButton.OK, MessageBoxImage.Error);
            }

            // Update the UI state if necessary
            UpdateUIState();
        }

        private void UpdateUIState()
        {
            // TODO: Implement logic to update UI state based on executed command
        }

        private void ShowCustomizationDialog()
        {
            // TODO: Implement a customization dialog for adding/removing Quick Access Toolbar items
        }
    }

    /// <summary>
    /// Represents an item in the Quick Access Toolbar
    /// </summary>
    internal class ToolbarItem
    {
        public string Label { get; }
        public ICommand Command { get; }
        public string IconPath { get; }

        public ToolbarItem(string label, ICommand command, string iconPath)
        {
            Label = label;
            Command = command;
            IconPath = iconPath;
        }
    }
}

// Human tasks:
// TODO: Implement a customization dialog for adding/removing Quick Access Toolbar items
// TODO: Add support for keyboard shortcuts for Quick Access Toolbar items
// TODO: Implement proper state management for toggleable toolbar items
// TODO: Add support for grouping related toolbar items
// TODO: Implement tooltips with extended information for toolbar items
// TODO: Add support for displaying recently used commands in the Quick Access Toolbar
// TODO: Implement synchronization of Quick Access Toolbar settings across devices
// TODO: Add support for exporting and importing Quick Access Toolbar configurations
// TODO: Implement accessibility features for screen readers
// TODO: Add support for high contrast themes in the Quick Access Toolbar
// TODO: Ensure proper scaling of toolbar for different DPI settings
// TODO: Implement customizable toolbar positioning (above/below the ribbon)
// TODO: Determine the most useful default items for the Quick Access Toolbar
// TODO: Implement localization for default item labels and tooltips
// TODO: Implement a user interface for managing custom Quick Access Toolbar items
// TODO: Ensure proper serialization and deserialization of custom toolbar items
// TODO: Implement drag-and-drop functionality for reordering toolbar items
// TODO: Add visual indicators for items that are also available in the ribbon
// TODO: Implement proper error handling for failed command executions
// TODO: Add support for command parameters in toolbar items